src/core/server/saved_objects/service/lib/repository.ts

```diff
@@ -997,6 +1181,54 @@ export class SavedObjectsRepository {
       }
     }
 
+    if (!disableExtensions && this._spacesExtension) {
+      try {
+        namespaces = await this._spacesExtension.getSearchableNamespaces(options.namespaces);
+      } catch (err) {
+        if (Boom.isBoom(err) && err.output.payload.statusCode === 403) {
+          // The user is not authorized to access any space, return an empty response.
+          return SavedObjectsUtils.createEmptyFindResponse<T, A>(options);
+        }
+        throw err;
+      }
+      if (namespaces.length === 0) {
+        // The user is authorized to access *at least one space*, but not any of the spaces they requested; return an empty response.
+        return SavedObjectsUtils.createEmptyFindResponse<T, A>(options);
+      }
+    }
+
+    // We have to first do a "pre-authorization" check so that we can construct the search DSL accordingly
+    const spacesToPreauthorize = new Set(namespaces);
+    let typeToNamespacesMap: Map<string, string[]> | undefined;
+    let preAuthorizationResult: CheckAuthorizationResult<'find'> | undefined;
+    if (!disableExtensions && this._securityExtension) {
+      preAuthorizationResult = await this._securityExtension.checkAuthorization({
+        types: new Set(types),
+        spaces: spacesToPreauthorize,
+        actions: ['find'],
+      });
+      if (preAuthorizationResult.status === 'unauthorized') {
+        // If the user is unauthorized to find *anything* they requested, return an empty response
+        this._securityExtension.addAuditEvent({
+          action: AuditAction.FIND,
+          error: new Error(`User is unauthorized for any requested types/spaces.`),
+          // TODO: include object type(s) that were requested?
+          // requestedTypes: types,
+          // requestedSpaces: namespaces,
+        });
+        return SavedObjectsUtils.createEmptyFindResponse<T, A>(options);
+      }
+      if (preAuthorizationResult.status === 'partially_authorized') {
+        typeToNamespacesMap = new Map<string, string[]>();
+        for (const [objType, entry] of preAuthorizationResult.typeMap) {
+          if (!entry.find) continue;
+          // This ensures that the query DSL can filter only for object types that the user is authorized to access for a given space
+          const { authorizedSpaces, isGloballyAuthorized } = entry.find;
```

Comment:
I think we're missing a test for this case within `repository.security_extension.test.ts`. Do you think it's wortwhile to add one?

Reply:
Good catch. I will add a test that executes find within 2 namespaces, and mock a checkAuthorization return simulating authz in only one of those spaces and validate the ES call is made with only the authorizes space. Is that what you had in mind?

Reply:
Yeah, I think that'll do the trick!